bundle/edu.gemini.lchquery.servlet/src/main/scala/edu/gemini/lchquery/servlet/LchQueryFunctor.scala

```diff
@@ -0,0 +1,230 @@
+        Option(prog.getProgramID).foreach(id => setReference(id.stringValue()))
+        setSemester(prog.semester.orNull)
+        setTitle(spProg.getTitle)
+        setContactScientistEmail(spProg.getContactPerson)
+        setNgoEmail(spProg.getNGOContactEmail)
+        setNotifyPi(spProg.getNotifyPi.displayValue)
+        setPiEmail(Option(spProg.getPIInfo).map(_.getEmail).orNull)
+        setRollover(spProg.getRolloverStatus.toYesNo.displayValue)
+
+        if (queryType != LchQueryFunctor.QueryType.ProgramQuery) {
+          setObservationsNode(new ObservationsNode() {
+            obsList.map { obs =>
+              new Observation() {
+                val spObs = obs.toSPObs.get
+                setAo(obs.ao.displayValue)
+                obs.instrumentType.foreach(setInstrument)
+                setName(spObs.getTitle)
+                Option(ObsClassService.lookupObsClass(obs)).foreach(c => setObsClass(c.displayValue))
+                setId(obs.getObservationID.stringValue)
+                setStatus(ObservationStatus.computeFor(obs).displayValue)
+                setTooPriority(Too.get(obs).getDisplayValue)
+
+                // Create the timing windows.
+                obs.timingWindows.filter(_.nonEmpty).foreach { tws =>
+                  setConditions(new Conditions() {
+                    setTimingWindowsNode(new TimingWindowsNode() {
+                      tws.map(makeTimingWindow).foreach(getTimingWindows.add)
+                    })
+                  })
+                }
+
+                if (queryType == LchQueryFunctor.QueryType.TargetQuery) {
+                  obs.targetEnvironment.foreach { env =>
+                    Option(env.getTargets) collect {
+                      case lst if lst.nonEmpty => lst.asScalaList
+                    } foreach { tgts =>
+                      setTargetsNode(new TargetsNode() {
+                        tgts.map(tgt => makeTargetNode(tgt, env)).foreach(getTargets.add)
+                      })
+                    }
+                  }
+
+                }
+              }
+            } foreach getObservations.add
+          })
+        }
+      }
+    }
+  }
+
+  def getResult: QueryResult = queryResult
+
+  override def mergeResults(functorCollection: util.Collection[IDBFunctor]): Unit = {
+    queryResult = new QueryResult() {
+      setProgramsNode(new ProgramsNode())
+    }
+    functorCollection.asScala.map(_.asInstanceOf[LchQueryFunctor]).foreach { f =>
+      queryResult.getProgramsNode.getPrograms.addAll(f.getResult.getProgramsNode.getPrograms)
+    }
+  }
+}
+
+object LchQueryFunctor {
+  private[LchQueryFunctor] val Log = Logger.getLogger(classOf[LchQueryFunctor].getName)
+  private[LchQueryFunctor] val MsPerSecond = 1000
+  private[LchQueryFunctor] val MsPerMinute = MsPerSecond * 60
+  private[LchQueryFunctor] val MsPerHour   = MsPerMinute * 60
+
+  private[LchQueryFunctor] val dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss z") {
```

Comment:
I'll look into it, if it seems appropriate. (I know pretty much nothing about it, but may as well learn about it at some point.)